enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,319 @@
+
+### Deletion Phase
+Describes when a machine has been marked for deletion but is still present
+in the API.  Various actions happen during this phase, such as draining a node,
+deleting an instance from a cloud provider, and deleting the node object.
+
+### Hook Implementing Controller (HIC)
+The Hook Implementing Controller describes a controller, other than the
+machine controller, that adds, removes, and/or responds to a particular
+lifecycle hook. Each lifecycle hook should have a single HIC, but an HIC
+can optionally manage one or more hooks.
+
+## Summary
+
+Defines a set of annotations that can be applied to a Machine which affect the
+linear progress of a Machine’s lifecycle after a Machine has been marked for
+deletion. These annotations are optional and may be applied during Machine
+creation, sometime after Machine creation by a user, or sometime after Machine
+creation by another controller or application.
+
+## Motivation
+
+Allow custom and 3rd party components to easily interact with a Machine or
+related resources while that Machine's reconciliation is temporarily paused.
+This pause in reconciliation will allow these custom components to take action
+after a Machine has been marked for deletion, but prior to the Machine being
+drained and/or associated instance terminated.
+
+In particular, this will allow the Etcd Operator to prevent a Control Plane
+Machine from being drained/removed until the Etcd Operator has had a chance
+to synchronise the Etcd data to the replacement Machine.
+
+This could also be used by other components such as Storage to allow them to
+ensure safe detachment of volumes before the Machine is deleted.
+
+### Goals
+
+- Define an initial set of hook points for the deletion phase.
+- Define an initial set and form of related annotations.
+- Define basic expectations for a controller or process that responds to a
+lifecycle hook.
+
+### Non-Goals
+
+- Create an exhaustive list of hooks; we can add more over time.
+- Create new machine phases.
+- Create a mechanism to signal what lifecycle point a machine is at currently.
+- Dictate implementation of controllers that respond to the hooks.
+- Implement ordering in the machine controller.
+- Require anyone to use these hooks for normal machine operations, these are
+strictly optional and for custom integrations only.
+
+## Proposal
+
+- Utilize annotations to implement lifecycle hooks.
+- Each lifecycle point can have 0 or more hooks.
+- Hooks do not enforce ordering.
+- Hooks found during machine reconciliation effectively pause reconciliation
+until all hooks for that lifecycle point are removed from a machine's annotations.
+
+### User Stories
+
+#### Story 1
+(pre-terminate) As an operator, I would like to have the ability to perform
+different actions between the time a Machine is marked deleted in the API and
+the time the Machine is deleted from the infrastructure provider.
+
+For example, when replacing a Control Plane Machine, ensure a new Control
```

Comment:
For this story If etcd operator is gaining access to edit machines anyway why do we need a new thing? couldn't this just be a finaliser?

Reply:
I'll add a note about finalizers to the alternatives section, but for now, my view on this is that finalizers are for the garbage collector. We could teach Machine API to view other finalizers as a block and prevent it from doing any removal until the other finalizers are removed, but this goes against the assumptions that Kubernetes has around finalizers and could break things further down the line.

IMO, it's better to have a specific and well defined way of doing this for Machines than to cram extra logic into a field like finalizers that wasn't really intended for this purpose

Reply:
One thing that isn't listed in those quotes, which I was always told should be the case for finalizers, is that they shouldn't imply ordering, do we believe that's a true statement? Because using finalizers for this approach does imply ordering.

The benefit of the new system to me is that we have something generic that works with multiple use cases.

If we were to make this a finalizer specific thing, I don't think it's safe to make the Machine API wait until all other finalizers are gone, for example, this would cause a deadlock between PVC finalizers and the Machine API.

That means then that we have to teach Machine API about a specific etcd finalizer to wait on, which feels like we now don't have a clear separation of concerns between the two projects and very much makes the MAO and ETCDO tied together for the future.

Having something generic like this ensures MAO doesn't need to know anything about the etcd operator, and allows extension in the future for other projects. It makes the relationship one way, rather than bi-directional, which IMO is a cleaner implementation

Reply:
Yeh that's the assumption for interoperability and the way I see this it would still be satisfied. To me our use case is the analogous to the one in the docs.
```
A common example of a finalizer is kubernetes.io/pv-protection, which prevents accidental deletion of PersistentVolume objects. When a PersistentVolume object is in use by a Pod, Kubernetes adds the pv-protection finalizer. If you try to delete the PersistentVolume, it enters a Terminating status, but the controller can't delete it because the finalizer exists. When the Pod stops using the PersistentVolume, Kubernetes clears the pv-protection finalizer, and the controller deletes the volume.
```

```
When a machine is master the etcd-protection finaliser is added. If you try to delete the Machine, it enters a Terminating status, but the controller can't delete it because the finalizer exists. When the Etcd stops using that Machine, the etcd-protection finalizer is cleared, and the controller deletes the machine.
```

The only order logic that we need is to enforce the etcd graceful removal is done before proceeding with the core infra deletion and can be implemented as above.
Isn't this otherwise actually proposing basically ordered finalisers but using an annotation instead? My concern is that's a loose public contract where any one can freely "petisize" any machine resulting in difficult to predict behaviour that will need to be maintained forever. Other than this I have no objections.